src/Agent/NewRelic/Agent/Core/Transactions/Transaction.cs

```diff
@@ -133,6 +133,11 @@ public void End(bool captureResponseTime = true)
                 {
                     LogFinest("Response time captured.");
                 }
+
+                // Once the response time is captured, the only work remaining for the transaction can be
+                // async work that is holding onto the transaction. We need to remove the transaction
+                // from the transaction context so that it will not be reused.
+                Agent._transactionService.RemoveOutstandingInternalTransactions(true, true);
```

Comment:
I think this may address some other issues, too, though I suspect we still have some async-tracking race conditions lurking. Good find!